collections/_posts/2022-07-07-C2FTrans-Segmentation.md

```diff
@@ -0,0 +1,141 @@
+
+*  $$ F_4 $$ is projected into $$ \{K_4, V_4\} $$, where $$ K_4, Q_4 \in \mathbb{R}^{\frac{\text{HW}}{64} \times \text{D}_h} $$.
+
+*  Likewise, $$ F_5 $$ is projected into $$ \{K_5, V_5\} $$, where $$ K_5, Q_5 \in \mathbb{R}^{\frac{\text{HW}}{256} \times \text{D}_h} $$.
+
+Cross-scale attention is then obtained by $$ F^i_{ca} (Q_{3,i}, K_i, V_i) = \text{softmax}(\frac{Q_{3,i} K^T_i}{\sqrt{\text{d}}}) V_i $$, where
+
+$$ F^4_{ca} \in (\mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}_h} \times \mathbb{R}^{\text{D}_h \times \frac{\text{HW}}{64}} \times \mathbb{R}^{\frac{\text{HW}}{64} \times \text{D}_h}) = \mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}_h} $$ 
+
+$$ F^5_{ca} \in (\mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}_h} \times \mathbb{R}^{\text{D}_h \times \frac{\text{HW}}{256}} \times \mathbb{R}^{\frac{\text{HW}}{256} \times \text{D}_h}) = \mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}_h} $$
+
+To be noted that during this step, multiscale features are mixed. According to the authors, this is crucial as lower resolution feature maps correspond to larger receptive fields, hence contain richer semantic information. By extracting $$ \text{K} $$ and $$ \text{V} $$ from lower resolution feature maps, computational complexity is reduced by at least a factor of 4 since their sequence length is shorter.  
+
+After that, $$ F^4_{ca} $$ and $$ F^5_{ca} $$ from all the transformer heads are concatenated and connected residually with $$ F^{'}_3 $$. To do so, another linear projection matrix $$ W_{ca} \in \mathbb{R}^{2 \text{gD}_h \times \text{D}} $$ is learned, where $$ \text{g} $$ represents the number of self-attention heads:
+
+$$
+\begin{align}
+
+    F_{ca} &= \underbrace{\text{concat}(F^4_{ca,1}, \cdots, F^4_{ca,g}, F^5_{ca,1}, \cdots, F^5_{ca,g})}_{\mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}_h \times \text{2g}} \xrightarrow[\text{reshape}]{} \mathbb{R}^{\frac{\text{HW}}{16} \times \text{2gD}_h}} \cdot \overbrace{W_{ca}}^{\mathbb{R}^{2 \text{gD}_h \times \text{D}}} + F^{'}_3 \in \mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}} \\
+
+    F_{ca} &= \text{FFN}(F_{ca}) + F^{'}_3  \\ 
+\end{align}
+$$
+
+To obtain the final CGT output, $$ F_{CGT} $$:
+> $$ F_{ca} \in \mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}} \xrightarrow[\text{reshape}]{} \mathbb{R}^{\text{D} \times \frac{\text{H}}{4} \times \frac{\text{W}}{4}} \xrightarrow[1 \times 1 \text{ conv}]{} F_{CGT} \in \mathbb{R}^{\text{D} \times \frac{\text{H}}{4} \times \frac{\text{W}}{4}} $$
+
+> ⚠️ <span style="color:red">In the paper, it is written that they learn two projection matrices and apply a residual connection to get $$ F_{CGT} $$, which contradicts their GitHub code.</span>
+
+Three functionalities of $$ F_{CGT} $$:
```

Comment:
From the text before: 

> To obtain the final CGT output, $$ F_{CGT} $$

I understand that F_CGT is the output. But in that case, I'm not sure I get what _functionalities_ an output can have? As in, what _functionalities_ can data have? Would it be more appropriate to talk about what transformations produced the output or downstream tasks? So could you please elaborate on this to make things clearer?